test/shared_utils/reference_hashing.py

```diff
@@ -0,0 +1,190 @@
+      timestamps reflecting when the output was created.
+    * For a variable, the array of data (values and shape).
+
+    """
+    metadata_bytes = get_metadata_bytes(xarray_object.attrs)
+    dimensions_bytes = get_dimensions_bytes(xarray_object.dims)
+
+    if isinstance(xarray_object, xr.Dataset):
+        variable_array_bytes = None
+    else:
+        variable_array_bytes = get_numpy_array_bytes(xarray_object.data)
+
+    return get_hash_value(
+        metadata_bytes,
+        dimensions_bytes,
+        variable_array_bytes,
+    )
+
+
+def get_hash_of_xarray_dataset(
+    dataset_path: str, dataset: xr.Dataset
+) -> dict[str, str]:
+    """Retrieve hashes of Dataset and all variables contained."""
+    return {
+        dataset_path: get_xarray_object_hash(dataset),
+        **{
+            get_full_variable_path(dataset_path, variable_path): get_xarray_object_hash(
+                variable
+            )
+            for variable_path, variable in dataset.variables.items()
+        },
+    }
+
+
+def get_full_variable_path(dataset_path: str, variable_path: str) -> str:
+    """Combine xarray.Dataset path and xarray.Variable path."""
+    return str(Path(dataset_path).joinpath(variable_path))
+
+
+def get_hash_value(
+    metadata_bytes: bytes,
+    dimension_bytes: bytes,
+    variable_array_bytes: bytes | None = None,
+):
+    """Return a string of a SHA 256 hash of the inputs.
+
+    This function can be used for both groups and
+    variables, with the final argument being optional.
+
+    """
+    all_bytes = metadata_bytes + dimension_bytes
+
+    if variable_array_bytes is not None:
+        all_bytes += variable_array_bytes
+
+    return sha256(all_bytes).hexdigest()
+
+
+def get_dimensions_bytes(variable_dimensions: tuple[str]) -> bytes:
+    """Convert xarray.DataArray.dims to bytes.
+
+    For variables with no dimensions, xarray.DataArray.dims is an
+    empty tuple.
+
+    """
+    return str(variable_dimensions).encode('utf-8')
+
+
+def get_numpy_array_bytes(numpy_array: np.ndarray) -> bytes:
+    """Convert a numpy array to a byte string.
+
+    The string is a combination of:
+
+    * The shape of the array.
+    * The values of each element of the array.
+
+    """
+    shape_byte_string = bytes(str(numpy_array.shape), encoding='utf-8')
+    array_byte_representation = numpy_array.tobytes()
+    return shape_byte_string + array_byte_representation
+
+
+def get_metadata_bytes(variable_metadata: dict) -> bytes:
+    """Convert xarray.DataArray.attrs to bytes.
+
+    * The attributes are filtered for values that will contain
+      datetimes reflecting processing times.
+    * The resulting dictionary will be sorted then converted to a string.
+    * That string will be encoded as UTF-8 bytes.
+
+    """
+    cleaned_metadata = {
+        key: serialise_metadata_value(value)
+        for key, value in variable_metadata.items()
+        if not is_varying_attribute(key)
```

Comment:
maybe I missed the correct location to ask about configuring varying metadata.  I suppose that can be a future feature